From 34758fcde60b934272415849940ee11da6328385 Mon Sep 17 00:00:00 2001
From: Vadim Pasternak <vadimp@nvidia.com>
Date: Thu, 2 Nov 2023 11:42:47 +0000
Subject: [PATCH v6.1 81/81] hwmon: (mlxreg-fan): Downstream: Allow fan speed
 setting granularity of 1 PWM

Currently PWM setting is allowed with 10 percent stepping.
Such configuration is aligned with thermal drivers, which are used to be
bound to "mlxreg-fan" driver.

This binding happens when the cooling instances created by the driver are
bound to some kernel thermal driver.

In case system is not using kernel thermal control and the cooling
instances created by the driver are not bound to any thermal driver, the
driver still does not allow setting of PWM granularity less than 10
percent.

Allow setting fan with one percent granularity, thus any user space
thermal application will be able to set PWM to any allowed value in range
from 51 PWM to 255 PWM.

Note: this is downstream patch, since it can affect functionality for
the Nvidia users running kernel thermal control. So, it is not going to be
submitted to up-stream.

Signed-off-by: Vadim Pasternak <vadimp@nvidia.com>
---
 drivers/hwmon/mlxreg-fan.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/hwmon/mlxreg-fan.c b/drivers/hwmon/mlxreg-fan.c
index 102a517940f3..1c4adcfd4498 100644
--- a/drivers/hwmon/mlxreg-fan.c
+++ b/drivers/hwmon/mlxreg-fan.c
@@ -15,10 +15,15 @@
 #define MLXREG_FAN_MAX_TACHO		24
 #define MLXREG_FAN_MAX_PWM		4
 #define MLXREG_FAN_PWM_NOT_CONNECTED	0xff
-#define MLXREG_FAN_MAX_STATE		10
+#ifdef CONFIG_MLXSW_CORE_THERMAL
+#define MLXREG_FAN_MAX_STATE            10
+#define MLXREG_FAN_SPEED_MIN_LEVEL      2       /* 20 percent */
+#else
+#define MLXREG_FAN_MAX_STATE            255
+#define MLXREG_FAN_SPEED_MIN_LEVEL      51       /* 20 percent */
+#endif
 #define MLXREG_FAN_MIN_DUTY		51	/* 20% */
 #define MLXREG_FAN_MAX_DUTY		255	/* 100% */
-#define MLXREG_FAN_SPEED_MIN_LEVEL		2	/* 20 percent */
 #define MLXREG_FAN_TACHO_SAMPLES_PER_PULSE_DEF	44
 #define MLXREG_FAN_TACHO_DIV_MIN		283
 #define MLXREG_FAN_TACHO_DIV_DEF		(MLXREG_FAN_TACHO_DIV_MIN * 4)
-- 
2.20.1

